### ISSUE TITLE
Build 64bit DLL for Unity(C#) and adding external libraries

### ISSUE THREAD

**Issue by ATHellboy** [libpd/libpd#201]
Hey guys,

I've tried to compile 64 bit version of libpd for Unity3D multiple times but Unity3D crashes after playing scene. I attached `out.text` file for executing `mingw64_build_csharp.bat` file.
[out.txt](https://github.com/libpd/libpd/files/1448715/out.txt)

Also I don't know how to include external libraries of PD in libpd. There is instruction for C language but I didn't find for C#. First, I want to use `freeverb~` in my patches.

**Comment by danomatika** [comment]
Most of the devs do not have an experience with Unity, so we cannot help much. There are a couple of libpd + Unity wrappers around but I'm not sure of their status. There were some threads on the pd-list that go into this: i.e. https://lists.puredata.info/pipermail/pd-list/2016-10/116777.html.

A couple points:

* Don't use the current master as it's most likely not stable due to some API changes in both the C core and prelim work in the C# layer, use a version tag
* Why compile at all when you can use the precompiled C# lib from NuGet: https://github.com/libpd/libpd#c

As for externals, libpd on desktop platforms can load compiled externals the same as Pd vanilla. You do not need to build the sources into your application. Easiest solution is to build free verb~ separately and have the loadable dll in your libpd search path.

**Comment by ATHellboy** [comment]
Thanks for replying.

I downloaded 64 bit version of libpd by NuGet and works like a charm.

For externals, I didn't figure out what exactly you said, if it is possible tell me which folder contains these externals. Also I use [uPD ](https://github.com/Magicolo/uPD).

**Comment by danomatika** [comment]
>I didn't figure out what exactly you said, if it is possible tell me which folder contains these externals

libpd can load pre-compiled externals (.so, .dll, etc) in the same way that Pd vanilla does, just put them in a place the loader can find them. On a desktop system, you do not need to build the externals' sources with libpd or your project, that's only required for systems like iOS which don't allow dynamic loading. One way to do this is to have the externals in the same folder as the main Pd patch. Another is to put them into a folder which you can add to libpd's search path (there is a function to add paths).

**Comment by danomatika** [comment]
Can C# + Windows people double check if the 64 bit build is still working? @residuum

**Comment by danomatika** [comment]
This issue might be related to #164. If you get a chance, could you pull the latest master, rebuilt for 64 bit using the mingw64 .bat file?

**Comment by ATHellboy** [comment]
I tested it there is a log: 

> make clobber
> rm -f pure-data/src/d_arithmetic.o pure-data/src/d_array.o pure-data/src/d_ctl.o pure-data/src/d_dac.o pure-data/src/d_delay.o pure-data/src/d_fft.o pure-data/src/d_fft_fftsg.o pure-data/src/d_filter.o pure-data/src/d_global.o pure-data/src/d_math.o pure-data/src/d_misc.o pure-data/src/d_osc.o pure-data/src/d_resample.o pure-data/src/d_soundfile.o pure-data/src/d_ugen.o pure-data/src/g_all_guis.o pure-data/src/g_array.o pure-data/src/g_bang.o pure-data/src/g_canvas.o pure-data/src/g_clone.o pure-data/src/g_editor.o pure-data/src/g_graph.o pure-data/src/g_guiconnect.o pure-data/src/g_hdial.o pure-data/src/g_hslider.o pure-data/src/g_io.o pure-data/src/g_mycanvas.o pure-data/src/g_numbox.o pure-data/src/g_readwrite.o pure-data/src/g_rtext.o pure-data/src/g_scalar.o pure-data/src/g_template.o pure-data/src/g_text.o pure-data/src/g_toggle.o pure-data/src/g_traversal.o pure-data/src/g_vdial.o pure-data/src/g_vslider.o pure-data/src/g_vumeter.o pure-data/src/m_atom.o pure-data/src/m_binbuf.o pure-data/src/m_class.o pure-data/src/m_conf.o pure-data/src/m_glob.o pure-data/src/m_memory.o pure-data/src/m_obj.o pure-data/src/m_pd.o pure-data/src/m_sched.o pure-data/src/s_audio.o pure-data/src/s_audio_dummy.o pure-data/src/s_inter.o pure-data/src/s_loader.o pure-data/src/s_main.o pure-data/src/s_path.o pure-data/src/s_print.o pure-data/src/s_utf8.o pure-data/src/x_acoustics.o pure-data/src/x_arithmetic.o pure-data/src/x_array.o pure-data/src/x_connective.o pure-data/src/x_gui.o pure-data/src/x_interface.o pure-data/src/x_list.o pure-data/src/x_midi.o pure-data/src/x_misc.o pure-data/src/x_net.o pure-data/src/x_scalar.o pure-data/src/x_text.o pure-data/src/x_time.o pure-data/src/x_vexp.o pure-data/src/x_vexp_if.o pure-data/src/x_vexp_fun.o libpd_wrapper/s_libpdmidi.o libpd_wrapper/x_libpdreceive.o libpd_wrapper/z_hooks.o libpd_wrapper/z_libpd.o pure-data/src/d_fft_fft_fftsg.o pure-data/src/d_fft_fftw.o pure-data/src/d_fft_fftsg_h.o pure-data/src/x_qlist.o libpd_wrapper/util/ringbuffer.o libpd_wrapper/util/z_queued.o jni/z_jni_plain.o
> rm -f  pure-data/extra/bob~/bob~.o pure-data/extra/bonk~/bonk~.o pure-data/extra/choice/choice.o pure-data/extra/fiddle~/fiddle~.o pure-data/extra/loop~/loop~.o pure-data/extra/lrshift~/lrshift~.o pure-data/extra/pique/pique.o pure-data/extra/sigmund~/sigmund~.o pure-data/extra/stdout/stdout.o
> rm -f libs/libpd.dll libs/libpdcsharp.dll java-build/org/puredata/core/natives/windows/x86_64//pdnative.dll libs/libpd.jar
> rm -f libs/`basename java-build/org/puredata/core/natives/windows/x86_64//pdnative.dll`
> rm -rf java-build
> 
> F:\Project\libpd>make csharplib -DPD_LONGINTTYPE="long long"
> make: invalid option -- D
> make: invalid option -- P
> make: invalid option -- D
> make: invalid option -- _
> Usage: make [options] [target] ...
> Options:
>   -b, -m                      Ignored for compatibility.
>   -B, --always-make           Unconditionally make all targets.
>   -C DIRECTORY, --directory=DIRECTORY
>                               Change to DIRECTORY before doing anything.
>   -d                          Print lots of debugging information.
>   --debug[=FLAGS]             Print various types of debugging information.
>   -e, --environment-overrides
>                               Environment variables override makefiles.
>   --eval=STRING               Evaluate STRING as a makefile statement.
>   -f FILE, --file=FILE, --makefile=FILE
>                               Read FILE as a makefile.
>   -h, --help                  Print this message and exit.
>   -i, --ignore-errors         Ignore errors from recipes.
>   -I DIRECTORY, --include-dir=DIRECTORY
>                               Search DIRECTORY for included makefiles.
>   -j [N], --jobs[=N]          Allow N jobs at once; infinite jobs with no arg.
>   -k, --keep-going            Keep going when some targets can't be made.
>   -l [N], --load-average[=N], --max-load[=N]
>                               Don't start multiple jobs unless load is below N.
>   -L, --check-symlink-times   Use the latest mtime between symlinks and target.
>   -n, --just-print, --dry-run, --recon
>                               Don't actually run any recipe; just print them.
>   -o FILE, --old-file=FILE, --assume-old=FILE
>                               Consider FILE to be very old and don't remake it.
>   -O[TYPE], --output-sync[=TYPE]
>                               Synchronize output of parallel jobs by TYPE.
>   -p, --print-data-base       Print make's internal database.
>   -q, --question              Run no recipe; exit status says if up to date.
>   -r, --no-builtin-rules      Disable the built-in implicit rules.
>   -R, --no-builtin-variables  Disable the built-in variable settings.
>   -s, --silent, --quiet       Don't echo recipes.
>   -S, --no-keep-going, --stop
>                               Turns off -k.
>   -t, --touch                 Touch targets instead of remaking them.
>   --trace                     Print tracing information.
>   -v, --version               Print the version number of make and exit.
>   -w, --print-directory       Print the current directory.
>   --no-print-directory        Turn off -w, even if it was turned on implicitly.
>   -W FILE, --what-if=FILE, --new-file=FILE, --assume-new=FILE
>                               Consider FILE to be infinitely new.
>   --warn-undefined-variables  Warn when an undefined variable is referenced.
> 
> This program built for i686-pc-msys
> Report bugs to <bug-make@gnu.org>

I need to say that I changed mingw64_build_csharp.bat a little. I changed MSYS directory from C:\msys64 to msys32 because there isn't any folder in MSYS by this name.

As I see, you added this line of code `make csharplib -DPD_LONGINTTYPE="long long"`, according to https://github.com/djkoloski/libpd64 I've changed `#define PD_LONGINTTYPE long` in m_pd.h to `#define PD_LONGINTTYPE long long int` before and didn't solve crash problem in Unity.

**Comment by ATHellboy** [comment]
Can you tell me what is the stable master of libpd for C# (Unity) ?

I have some problems by 64 bit version that I downloaded from NuGet, my scene crashes whenever I play Sequence scene of [uPD](https://github.com/Magicolo/uPD) but other scenes works good. There isn't any problem with 32 bit version in all the scenes.

**Comment by ATHellboy** [comment]
Also I've not figured out yet how to use pre-compiled externals in my project. Can you tell where is the folder for main Pd patches and How can I add path for loader to load externals ?

**Comment by ATHellboy** [comment]
Accidentally I've closed this issue and didn't know that.

**Comment by ATHellboy** [comment]
Anyone ?

**Comment by danomatika** [comment]
>Can you tell me what is the stable master of libpd for C# (Unity) ?

I guess there currently isn't one as you're finding out. (?) You could try a release tag but no guarantee.

>I have some problems by 64 bit version that I downloaded from NuGet, my scene crashes whenever I play Sequence scene of uPD but other scenes works good. There isn't any problem with 32 bit version in all the scenes.

The libpd developers don't use Unity directly. As such we can't really off much help beyond getting the C# wrapper to build and it does not provide any Unity-oriented features.. As for using it within Unity, we simply don't know. If uPD works, why not use that instead of creating your own as it is probably using the C# wrapper anyway.

>Also I've not figured out yet how to use pre-compiled externals in my project. Can you tell where is the folder for main Pd patches and How can I add path for loader to load externals ?

I already mentioned this in a previous post: you need to place the externals in a place libpd can find them and possible add to libpd's search paths using the applicable function in the libpd API.

All applications start with a working directory and libpd uses this by default. Sometimes it's the home directory or within a mac application bundle but it's usually the directory where the application binary sits. It's really up to how the program starts and/or modifies it's path after it starts up.

I have no idea how Unity handles it's working path. That's something you need to look into. Maybe it starts and the sets the working path to a folder where it stores it's data, etc? In any case, once you know what the working path is, you can either place your patches and externals there or change it.

**Comment by ATHellboy** [comment]
> If uPD works, why not use that instead of creating your own as it is probably using the C# wrapper anyway.
Unfortunately, It is 32 bit version of libpd and it is for almost 2 years ago.

**Comment by danomatika** [comment]
>Unfortunately, It is 32 bit version of libpd and it is for almost 3 years ago.

Hmm, then I suggest opening an issue with that project. As far as I know, the C# libpd wrapper needs some extra code to adapt it to work with Unity which is why there are several different libpd for Unity projects. I'm not sure we (the main libpd devs) have enough Unity knowledge overall.

**Comment by danomatika** [comment]
@ATHellboy If you have a chance, can you try pulling the current master and rebuilding. @residium might have fixed my fix from a week ago...

**Comment by danomatika** [comment]
@ATHellboy If you have a chance, can you try pulling the current master and rebuilding. @residium might have fixed my fix from a week ago...

**Comment by anthonyburchell** [comment]
@danomatika im working on solving this exact problem. I have tried the new bat file and it failed on the first run. (or rather hung on the cmd window without closing.) I ran it again and it worked nicely! However with some modification to the bat file. Currently in the process of getting it back into unity to test osc communication. In the readme instructions for build instructions it says that using the 32bit version was the way to go because it can build for 64 as well. It might be helpful to add instructions to edit the bat file to reference the correct version of msys.  I got stuck on that for a bit haha. I'll report back if I have success in this :) so far so good

**Comment by anthonyburchell** [comment]
well that crashed when i loaded into unity ^_^ going to try installing 64bit msys and rebuild through that.

**Comment by residuum** [comment]
I have changed that wrong advice from the README.

**Comment by anthonyburchell** [comment]
I made sure to include the native scripts in the unity project (something I had not done before) and I've made a tiny bit of progress. In the PureDataFilterRead.cs file there is an error throw that reads: 
`Assets/Magicolo/AudioTools/PureData/PureDataFilterRead.cs(44,11): error CS1502: The best overloaded method match for `LibPDBinding.LibPD.Process(int, short[], short[])' has some invalid arguments
`
and also: 
`Assets/Magicolo/AudioTools/PureData/PureDataFilterRead.cs(44,42): error CS1503: Argument `#2' cannot convert `System.IntPtr' expression to type `short[]'
`

The lines throwing the invalid args is:
```
			if (pureData.bridge.initialized && focused && !paused && !pureData.editorHelper.editorPaused) {
				LibPD.Process(pureData.bridge.ticks, dataPtr, dataPtr);
			}


```

If it is getting to the point of failing on invalid arguments to libpd is it safe to say that the rebuild of libpdcsharp.dll is loading correctly now and the issue is just that these other scripts in the project are crashing? 

hopefully that's helpful. I'm a noob working with this library :)

**Comment by anthonyburchell** [comment]
@ATHellboy fwiw I found a working 64bit dll here: https://github.com/djkoloski/libpd64 make sure to grab the native scripts from the libpd folder. Just keep in mind this is from 2014 and a lot of stuff is missing.

I still want to get a somewhat updated version so I can use the OSC message stuff added in the 2016 version of pd.

**Comment by ATHellboy** [comment]
> @ATHellboy If you have a chance, can you try pulling the current master and rebuilding. @residuum might have fixed my fix from a week ago...

I built it, but no chance also there is a problem in mingw64_build_csharp.bat file, because there isn't any msys64 folder, I tried to change it into MSYS=C:\msys32 but not sure if it is OK or not.

> @ATHellboy fwiw I found a working 64bit dll here: https://github.com/djkoloski/libpd64 make sure to grab the native scripts from the libpd folder. Just keep in mind this is from 2014 and a lot of stuff is missing.
> 
> I still want to get a somewhat updated version so I can use the OSC message stuff added in the 2016 version of pd.

I found it before and just I did his advice and fixes and I don't know why I didn't go to the folders and use his 64bit build. It's working like a charm now although it is a little old.

Do you have any idea where is the default path for loading external libraries ?

**Comment by ATHellboy** [comment]
Another question about `libwinpthread-1.dll` file. What is this file ? and is it possible to build it ? because when I build, it doesn't add new `libwinpthread-1.dll`.

**Comment by residuum** [comment]
Your line `LibPD.Process(pureData.bridge.ticks, dataPtr, dataPtr);`, because you write to the same location as your input. This is not the intended way.

About `libwinpthread-1.dll`: This library maps Windows threading to POSIX threads. It is not built by the Makefile, but must be there. You must include the correct version in your library.

**Comment by ATHellboy** [comment]
@residuum Do you have any idea where is the default path for loading external libraries ?

**Comment by residuum** [comment]
https://msdn.microsoft.com/en-us/library/windows/desktop/ms682586(v=vs.85).aspx

**Comment by ATHellboy** [comment]
Hey,

I pulled last changes and replaced casharp native scripts, but there is a error for Process in libpd, in these lines:

```
void OnAudioFilterRead(float[] data, int channels) {
			if (dataPtr == IntPtr.Zero) {
				dataHandle = GCHandle.Alloc(data, GCHandleType.Pinned);
				dataPtr = dataHandle.AddrOfPinnedObject();
			}
			if (pureData.bridge.initialized && focused && !paused && !pureData.editorHelper.editorPaused) {
				LibPD.Process(pureData.bridge.ticks, dataPtr, dataPtr);
			}
```

it is because of `datatptr` type, you changes it from `IntPtr` to `short[]`, for these lines I don't know how to get `dataPtr` for sending to `Process `function of libpd.  Any advice ?

**Comment by residuum** [comment]
Both overloads are still there:
https://github.com/libpd/libpd/blob/master/csharp/Native/LibPDNativeMethods.cs#L343
https://github.com/libpd/libpd/blob/master/csharp/Native/LibPDNativeMethods.cs#L414

I have not found a version, where `LibPD.Process()` had a overload for using `IntPtr` (if this is the type of `dataPtr`). Try casting:

    LibPd.Process(puredata.bridge.ticks, (float *)dataPtr, (float *)dataPtr);

There is also an overload that takes `float[]` directly. This may also work:

    void OnAudioFilterRead(float[] data, int channels) {
        if (pureData.bridge.initialized && focused && !paused && !pureData.editorHelper.editorPaused) {
            LibPD.Process(pureData.bridge.ticks, data, data);
        }
    }

**Comment by residuum** [comment]
Both overloads are still there:
https://github.com/libpd/libpd/blob/master/csharp/Native/LibPDNativeMethods.cs#L343
https://github.com/libpd/libpd/blob/master/csharp/Native/LibPDNativeMethods.cs#L414

I have not found a version, where `LibPD.Process()` had a overload for using `IntPtr` (if this is the type of `dataPtr`). Try casting:

    LibPd.Process(puredata.bridge.ticks, (float *)dataPtr, (float *)dataPtr);

There is also an overload that takes `float[]` directly. This may also work:

    void OnAudioFilterRead(float[] data, int channels) {
        if (pureData.bridge.initialized && focused && !paused && !pureData.editorHelper.editorPaused) {
            LibPD.Process(pureData.bridge.ticks, data, data);
        }
    }

**Comment by ATHellboy** [comment]
Oh, I checked the uPD `LibPDNativeMethods.cs` script and I think the developer has written it by himself before, because there is a overload function of Process with IntPtr input with `EntryPoint=libpd_process_unity`:

```
[DllImport("libpdcsharp", EntryPoint="libpd_process_unity")]
		private static extern  int process_unity(int ticks, [In] IntPtr inBuffer, [Out] IntPtr outBuffer, ref double time) ;
		[MethodImpl(MethodImplOptions.Synchronized)]
		public static int Process(int ticks, IntPtr inBuffer, IntPtr outBuffer, ref double time)
		{
			return process_unity(ticks, inBuffer, outBuffer, ref time);
		}
```

**Comment by residuum** [comment]
So, not a bug in libPd. Is the rest working?

**Comment by ATHellboy** [comment]
@residuum not sure about that, but Unity crashed before, by changing Process function in script it just now freezes. for now I don't know what to do and what Process function do.

**Comment by ATHellboy** [comment]
I've not figure out yet how to use external libraries ? I've tested many things but nothing happened.

**Comment by danomatika** [comment]
Any update on this? If it's not a libpd issue in the end, can you please close this?

**Comment by ATHellboy** [comment]
At that moment, I used previous version of the C# DLL and it's good but it doesn't have latest features.

I didn't check newer version if there is.

And I leaved importing external libraries for now because I haven't found out how it is working. I'm working on something else by libpd in Unity but still seeking the way for importing external libraries and have latest version of this library in Unity, so not sure if it is good idea to close this issue or not :)
